### PR TITLE
[util, dv] Auto-generate and add testplan for security countermeasures

### PR DESCRIPTION
**Pull request by sriyerg**
The first commit updates the reggen tool to parse the countermeasures and add a new testplan for it if not already created. 

The second commit uses the reggen tool updates to create the required testplans for all IPs.

### CODE REVIEW

**Comment by sriyerg** [comment]
TODOs:

- [ ] Fix descriptions in the generated text
- [ ] Deal with templated IPs (hw/ip_templates/*)

File: util/reggen/gen_sec_cm_testplan.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    outfile = Path(outdir) / f"{block.name.lower()}_sec_cm_testplan.hjson"
+    if outfile.exists():
+        names_from_testplan = []
+        with open(outfile, "r", encoding='UTF-8') as f:
+            data = hjson.load(f)
+            try:
+                names_from_testplan = [tp["name"] for tp in data["testpoints"]]
+            except KeyError as e:
+                log.fatal(f"Malformed testplan {outfile}")
+                return 1
```
Comment (msfschaffner) [inline util/reggen/gen_sec_cm_testplan.py:35]: I believe we've started to raise errors in other places of reggen instead of using `log.error` or `log.fatal`. E.g. https://github.com/lowRISC/opentitan/blob/91ceb766586bc98892ba8cf0cbe2aae9a13caa28/util/reggen/ip_block.py#L179-L205 I know it's a detail - just mentioning that there is an opportunity to align.

File: hw/ip/aon_timer/data/aon_timer_testplan.hjson [hunk 2]
```diff
@@ -76,7 +77,7 @@
       desc: '''
         Includes possible values of all registers of AON timer. For 32b registers, bins are
         introduced in order to simplfy coverage. Maximum and minimum values are collected in
-        a seperate bin to guarantee that they are checked. 32 separate bins are auto-generated 
+        a seperate bin to guarantee that they are checked. 32 separate bins are auto-generated
```
Comment (msfschaffner) [inline hw/ip/aon_timer/data/aon_timer_testplan.hjson:80]: nit: somewhat unrelated - but since we;re touching the file anyways: seperate -> separate

File: hw/ip/lc_ctrl/data/lc_ctrl_sec_cm_testplan.hjson [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    }
+    {
+      name: sec_cm_manuf_state_bkgn_chk
+      desc: '''
+            Verifies the countermeasure MANUF.STATE.BKGN_CHK.
+
+            Countermeasure description:
+            The manufacturing state vector is continuously decoded and checked,
+once the life cycle controller has initialized. If any mismatch is detected,
+local escalation is triggered (MAIN.FSM.LOCAL_ESC).
```
Comment (msfschaffner) [inline hw/ip/lc_ctrl/data/lc_ctrl_sec_cm_testplan.hjson:84]: Should we fix the template to handle the indents here automatically, or are we fine with editing this by hand (since they presumably need to be updated for V2S anyways)?

File: hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson [hunk 4]
```diff
@@ -41,15 +41,15 @@
...
             Verify the combo detection with ec_rst action.
-                                     
-            * Trigger the input keys by configuring COM_SEL_CTL_0 register.                                            
-            * Set the combo duration via the COM_DET_CTL_0 register.                                                   
-            * Select the action to be taken by configuring COM_OUT_CTL_0 register.                                     
-            * Set the pulse width via EC_RST_CTL register only to raise ec_rst action.                                 
-            * Read the COMBO_INTR_STATUS register to check if the interrupt is raised and 
-              clear the interrupt. 
-            * NOTE: This is a directed test with no random values for V1 milestone, 
-              further this test will be randomized.
```
Comment (msfschaffner) [inline hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson:52]: nit: looks like these are unrelated formatting changes. would be great to have them in a separate commit (just to be able to tell them appart). no big deal for this PR, though.

**Review by msfschaffner — COMMENTED** [review]
Thanks for adding this! It looks great overall, just a few minor comments.

File: hw/ip/adc_ctrl/data/adc_ctrl_sec_cm_testplan.hjson [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+// This testplan is auto-generated only the first time it is created. This is
+// because this testplan needs to be hand-editable. It is possible that these
+// testpoints can go out of date if the spec is updated with new
+// countermeasures. When `reggen` is invoked when this testplan already exists,
+// It checks if the list of testpoints is up-to-date and enforces the user to
+// make further manual updates.
+{
+  testpoints: [
+    {
+      name: sec_cm_bus_integrity
```
Comment (weicaiyang) [inline hw/ip/adc_ctrl/data/adc_ctrl_sec_cm_testplan.hjson:16]: this countermeasure is covered by the testpoint below. not sure how we will define the relationship between these 2 testpints?  
https://github.com/lowRISC/opentitan/blob/199d45626f8a7ae2aef5d9ff73793bf9a4233711/hw/dv/tools/dvsim/testplans/tl_device_access_types_testplan.hjson#L58-L65

File: hw/ip/adc_ctrl/data/adc_ctrl_sec_cm_testplan.hjson [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+      desc: '''
+            Verifies the countermeasure BUS.INTEGRITY.
+
+            Countermeasure description:
+            End-to-end bus integrity scheme.
```
Comment (weicaiyang) [inline hw/ip/adc_ctrl/data/adc_ctrl_sec_cm_testplan.hjson:21]: i think the `desc` is meant to describe what to test and how. Replicating the design description here seems a bit redundant and may bring more work for maintenance. 

If we want to show both design description and test description, we can show them in the rendered document. For hjson, we'd better put the necessary info. For example, we could put the testpoint along with countermeasures, and describe the detail of the testpoint in testplan. In rendered DV DOC, we can show countermeasure description, testpoint and testpoint description all together.

```
  countermeasures: [
    { name: "BUS.INTEGRITY",
      desc: "End-to-end bus integrity scheme."
      testpoint: "tl_intg_err"
    }

```

**Comment by cindychip** [comment]
Thanks Sri, Could you also help auto-generate the ones under ip_template? like alert_handler?

**Comment by sriyerg** [comment]
> Thanks Sri, Could you also help auto-generate the ones under ip_template? like alert_handler?

Yes  - see my second TODO in the above comment :)

File: hw/ip/adc_ctrl/data/adc_ctrl_sec_cm_testplan.hjson (hunk 5 above)
Comment (sriyerg) [reply hw/ip/adc_ctrl/data/adc_ctrl_sec_cm_testplan.hjson:16]: IMO duplication is fine (as we discussed offline) - user just needs to wire up the test `"{name}_tl_intg_err"` to this testpoint as well, which is a very small extra work.

File: hw/ip/aon_timer/data/aon_timer_testplan.hjson (hunk 2 above)
Comment (sriyerg) [reply hw/ip/aon_timer/data/aon_timer_testplan.hjson]: Fixed.

File: util/reggen/gen_sec_cm_testplan.py (hunk 1 above)
Comment (sriyerg) [reply util/reggen/gen_sec_cm_testplan.py]: Done.

**Review by weicaiyang — APPROVED** [review]
Looks great!

File: hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson (hunk 4 above)
Comment (msfschaffner) [reply hw/ip/sysrst_ctrl/data/sysrst_ctrl_testplan.hjson:52]: oh I see. yeah nevermind then.

**Review by msfschaffner — APPROVED** [review]

**Comment by sriyerg** [comment]
> > Thanks Sri, Could you also help auto-generate the ones under ip_template? like alert_handler?
> 
> Yes - see my second TODO in the above comment :)

This work is now complete - please re-review, thanks!

File: hw/top_earlgrey/ip/pwrmgr/data/autogen/pwrmgr_sec_cm_testplan.hjson [hunk 7]
```diff
@@ -0,0 +1,39 @@
+// Copyright lowRISC contributors.
```
Comment (tjaychen) [inline hw/top_earlgrey/ip/pwrmgr/data/autogen/pwrmgr_sec_cm_testplan.hjson:1]: quick question, i thought @matutem kept most of his test plans under `hw/ip/pwrmgr`, should those be moved here also?

Comment (tjaychen) [reply hw/top_earlgrey/ip/pwrmgr/data/autogen/pwrmgr_sec_cm_testplan.hjson:1]: a lot of this is because the weird templating... but i should work on converting this to ip_autogen and all will be clearer.

Comment (sriyerg) [reply hw/top_earlgrey/ip/pwrmgr/data/autogen/pwrmgr_sec_cm_testplan.hjson:1]: Yes. This is a duplicate of `hw/ip/<IP>/data/<IP>_sec_cm_testplan.hjson` which is also generated in this PR, where <IP>  is the list of IPs that are generated themselves. These toplevel specific ones are generated by `topgen` based on `<IP>.hjson.tpl` template, which I consider to be more accurate than the vanilla `<IP>.hjson`. See the vanilla `clkmgr.hjson` - it does not even list any countermeasures (but its `.tpl` counterpart does.

File: hw/ip/pinmux/data/pinmux_sec_cm_testplan.hjson [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+// These countermeasures and their descriptions can be found here:
+// .../pinmux/data/pinmux.hjson
+//
+// It is possible that the testing of some of these countermeasures may already
+// be covered as a testpoint in a different testplan. This duplication is ok -
+// the test would have likely already been developed. We simply map those tests
+// to the testpoints below using the `tests` key.
+//
+// Please ensure that this testplan is imported in:
+// .../pinmux/data/pinmux_testplan.hjson
```
Comment (matutem) [inline hw/ip/pinmux/data/pinmux_sec_cm_testplan.hjson:23]: Is this file used? Are we only using the corresponding autogen file?

File: hw/ip/pwrmgr/data/pwrmgr_testplan.hjson [hunk 9]
```diff
@@ -7,7 +7,10 @@
   import_testplans: ["hw/dv/tools/dvsim/testplans/csr_testplan.hjson",
                      "hw/dv/tools/dvsim/testplans/intr_test_testplan.hjson",
                      "hw/dv/tools/dvsim/testplans/tl_device_access_types_testplan.hjson",
-                     "hw/dv/tools/dvsim/testplans/stress_all_with_reset_testplan.hjson"]
+                     "hw/dv/tools/dvsim/testplans/stress_all_with_reset_testplan.hjson",
+                     // TODO: Top-level specific Hjson imported here. This will likely be resolved
+                     // once we move to IPgen flow.
+                     "hw/top_earlgrey/ip/pwrmgr/data/autogen/pwrmgr_sec_cm_testplan.hjson"]
```
Comment (matutem) [inline hw/ip/pwrmgr/data/pwrmgr_testplan.hjson:13]: The pwrmgr_testplan.hjson file will need to become specific to top_earlgrey at some point.

File: hw/ip/trial1/data/trial1_sec_cm_testplan.hjson [hunk 10]
```diff
@@ -0,0 +1,16 @@
+// Copyright lowRISC contributors.
```
Comment (matutem) [inline hw/ip/trial1/data/trial1_sec_cm_testplan.hjson:1]: Do we need this file?

File: util/regtool.py [hunk 11]
```diff
@@ -11,10 +11,10 @@
 import sys
 from pathlib import Path
 
-from reggen import (gen_cheader, gen_dv, gen_fpv, gen_html,
-                    gen_json, gen_rtl, gen_rust, gen_selfdoc, version)
-from reggen.ip_block import IpBlock
+from reggen import (gen_cheader, gen_dv, gen_fpv, gen_html, gen_json, gen_rtl,
+                    gen_rust, gen_sec_cm_testplan, gen_selfdoc, version)
 from reggen.countermeasure import CounterMeasure
```
Comment (matutem) [inline util/regtool.py:16]: Funny, I thought imports were ordered alphabetically, so I don't understand this change.

File: hw/top_earlgrey/ip/pwrmgr/data/autogen/pwrmgr_sec_cm_testplan.hjson (hunk 7 above)
Comment (matutem) [reply hw/top_earlgrey/ip/pwrmgr/data/autogen/pwrmgr_sec_cm_testplan.hjson:1]: Right, this whole autogen business needs reshuffling... later...

**Review by matutem — APPROVED** [review]
With a few questions and suggestions, LGTM

File: hw/ip/csrng/data/csrng_testplan.hjson [hunk 12]
```diff
@@ -7,13 +7,14 @@
...
+                     "hw/dv/tools/dvsim/testplans/stress_all_with_reset_testplan.hjson",
+                     "csrng_sec_cm_testplan.hjson"]
   testpoints: [
     {
       name: smoke
       desc: '''
-	    Verify sending instantiate/generate cmds via SW path.
-	    Verify reading genbits via SW path.
+      Verify sending instantiate/generate cmds via SW path.
+      Verify reading genbits via SW path.
```
[the remark this answers is not in the document]
Comment (sriyerg) [reply hw/ip/csrng/data/csrng_testplan.hjson:17]: Hmm, these lines got inadvertently modified because I have a "replace preceding tabs with spaces on save" macro enabled on my vim. Perhaps ok to let this slide?

File: hw/ip/pinmux/data/pinmux_sec_cm_testplan.hjson (hunk 8 above)
Comment (sriyerg) [reply hw/ip/pinmux/data/pinmux_sec_cm_testplan.hjson:23]: This autogenerated testplan is not used (and the file referenced in the line above does not exist either), but we will need to check it in regardless, otherwise it will show up as a untracked file when people run `make -C hw`, which would be annoying.

File: hw/ip/pwrmgr/data/pwrmgr_testplan.hjson (hunk 9 above)
Comment (sriyerg) [reply hw/ip/pwrmgr/data/pwrmgr_testplan.hjson:13]: Should be cleaned up once pwrmgr is pulled into IPgen flow.

File: hw/ip/trial1/data/trial1_sec_cm_testplan.hjson (hunk 10 above)
Comment (sriyerg) [reply hw/ip/trial1/data/trial1_sec_cm_testplan.hjson:1]: Same comment as earlier - if we dont check it in, it will show up as untracked file which would be annoying.

File: util/regtool.py (hunk 11 above)
Comment (sriyerg) [reply util/regtool.py:16]: `reggen.c...` comes before `reggen.i...`. This (and few other unrelated formatting changes below) were done by `isort` and `yapf` so I did not bother questioning them :)

File: hw/ip/aon_timer/data/aon_timer_testplan.hjson [hunk 13]
```diff
@@ -76,7 +77,7 @@
       desc: '''
         Includes possible values of all registers of AON timer. For 32b registers, bins are
         introduced in order to simplfy coverage. Maximum and minimum values are collected in
-        a seperate bin to guarantee that they are checked. 32 separate bins are auto-generated 
+        a separate bin to guarantee that they are checked. 32 separate bins are auto-generated
         in order to distrubute values evenly.
```
[the remark this answers is not in the document]
Comment (sriyerg) [reply hw/ip/aon_timer/data/aon_timer_testplan.hjson:81]: This should probably be fixed in a separate PR.

File: hw/top_earlgrey/ip/pwrmgr/data/autogen/pwrmgr_sec_cm_testplan.hjson (hunk 7 above)
Comment (tjaychen) [reply hw/top_earlgrey/ip/pwrmgr/data/autogen/pwrmgr_sec_cm_testplan.hjson:1]: sounds good. I think `pwr/rst/clk`mgrs really need to be shifted to `ip_autogen` so there is less ambiguity overall.
I will...do that....some day.

**Review by tjaychen — APPROVED** [review]

File: hw/ip/csrng/data/csrng_testplan.hjson (hunk 12 above)
[the remark this answers is not in the document]
Comment (matutem) [reply hw/ip/csrng/data/csrng_testplan.hjson:17]: If you have to...

File: hw/ip/pinmux/data/pinmux_sec_cm_testplan.hjson (hunk 8 above)
Comment (matutem) [reply hw/ip/pinmux/data/pinmux_sec_cm_testplan.hjson:23]: Ack

File: hw/ip/trial1/data/trial1_sec_cm_testplan.hjson (hunk 10 above)
Comment (matutem) [reply hw/ip/trial1/data/trial1_sec_cm_testplan.hjson:1]: Gee, I never noticed there was a trial1 ip :)

File: util/regtool.py (hunk 11 above)
Comment (matutem) [reply util/regtool.py:16]: My goof... it is alphabetical now :)

**Review by matutem — APPROVED** [review]
LGTM

**Comment by sriyerg** [comment]
Merging this in, thanks for the comments!
